core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -209,7 +211,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
    */
   private val replayExecutor: ExecutorService = {
     if (!Utils.isTesting) {
-      ThreadUtils.newDaemonFixedThreadPool(NUM_PROCESSING_THREADS, "log-replay-executor")
+      BlockingThreadPoolExecutorService.newInstance(
+        numReplayThreads, 1024, 60L, TimeUnit.SECONDS, "log-replay-executor")
```

Comment:
1024 - queue size, since listing logDir may take some times, it's better to allow some tasks to be queued before starting the next round
60s, core thread idle time

I don't think it's necessary to expose these configs to users, but I'm open to different ideas.